rgw/restore: Fixed status codes and response headers

shreyanshjain7174 opened the pull request:
Added x-amz-restore header and fixed status codes for both restore-cli
and read_through 

Fixed the list objects : Storage class metavars for temporary restore.

Signed-off-by: shreyanshjain7174 <ssanchet@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

pereman2 committed:
qa/rados/upgrade: ignore PG_DEGRADED

Fixes: https://tracker.ceph.com/issues/67182
Signed-off-by: Pere Diaz Bou <pere-altea@hotmail.com>

Commit:
rgw/restore: Fixed status codes and response headers

Added x-amz-restore header and fixed status codes for both restore-cli
and read_through

Signed-off-by: shreyanshjain7174 <ssanchet@redhat.com>

thotz commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +467,33 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+
+  if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::RestoreFailed) {
+    s->err.http_ret = 202; //Accepted
+    dump_header(s, "x-amz-restore", "on-going-request=\"true\"");
```
should we treat restore failure as error than 202? And the on-going-request should be true only for restoreinprogress scenario, all other cases it should be false, right?

shreyanshjain7174 replied:
as per aws docs, only the cloud restored request had on-going request as false with update expiry date.

thotz commented on src/rgw/rgw_op.cc:
```diff
@@ -5280,28 +5280,13 @@ void RGWRestoreObj::execute(optional_yield y)
+    ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
+    <<". Failing with " << op_ret << dendl;
+    if (op_ret == -ERR_INVALID_OBJECT_STATE) {
+      s->err.message = "This object was transitioned to cloud-s3";
```
error message here need to modified as well

soumyakoduri replied:
`s->err.message`  already seems to be set in handle_cloudtier_obj

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +467,33 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+    s->err.http_ret = 202; //Accepted
+    dump_header(s, "x-amz-restore", "on-going-request=\"true\"");
+  } else if (restore_status == rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
+    s->err.http_ret = 409; // Conflict
```
AWS specs mention about response codes only for Restore Object CLI but not for get-object/read-through. Lets not update http_ret in this routine. 

In addition, we ideally should set `x-amz-restore` only for HEAD request. Can you check if there is any way (might be there) to verify its HEAD request and then set `x-amz-restore` .

shreyanshjain7174 replied:
Okay sure. Will update that then.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3525,26 +3557,47 @@ void RGWRestoreObj_ObjStore_S3::send_response()
 
   rgw::sal::Attrs attrs = s->object->get_attrs();
   auto attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
+  auto expire_iter = attrs.find(RGW_ATTR_RESTORE_EXPIRY_DATE);
```
Probably its good to read & save these attrs in RGWRestoreObj::execute() before calling handle_cloudtier_obj(). That way we can identify if its the first restore request as attrs will not be set by then.

shreyanshjain7174 replied:
Makes sense! will try that out.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3525,26 +3557,47 @@ void RGWRestoreObj_ObjStore_S3::send_response()
+  } else if (restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
+    if (rt == rgw::sal::RGWRestoreType::Temporary) {
+    s->err.http_ret = 200; // OK temporary restore will have modified expiry date.
+    dump_header(s, "x-amz-restore", "ongoing-request=\"false\", expiry-date=\""+ dump_time_to_str(expiration_date) +"\"");
```
expiry_date attr is stored in string format. We need to convert it to real_time before calling dump_time_to_str().. refer to "parse_rgwx_mtime".

shreyanshjain7174 replied:
Will check.

cyx1231st committed:
crimson/os/seastore: misc cleanups

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

cyx1231st committed:
crimson/os/seastore: convert length logs to the hex format

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

cyx1231st committed:
crimson/os/seastore/transaction_manager: misc cleanups

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

cyx1231st committed:
crimson/os/seastore/transaction_manager: refine logs

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

myoungwon committed:
crimson/os/seastore: fix data inconsistency during ool writes

In RBM, seastore issues ool writes with allocated address.
If a transaction conflict occurs at this point, the allocated address is freed,
allowing the address to be reused.
However, data inconsistency can occur if seastore issues ool writes with
freed address before the preceding ool write has not been complete.

To fix this issue, this commit frees the allocated address after ool writes is
don in the event of the transaction conflict after ool write is issued.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

thotz commented on src/rgw/rgw_op.cc:
```diff
@@ -5279,29 +5279,38 @@ void RGWRestoreObj::execute(optional_yield y)
   }
-  rgw::sal::Attrs attrs = s->object->get_attrs();
-  auto attr_iter = attrs.find(RGW_ATTR_MANIFEST);
+  attrs = s->object->get_attrs();
```
I feel all the changes in this function need to filled post calling handle_cloud_tier object, then only those will be filled properly. Do we need to fill this here, it is overwritten in send_reponse right?

shreyanshjain7174 replied:
As per soumya 
>>> Probably its good to read & save these attrs in RGWRestoreObj::execute() before calling handle_cloudtier_obj(). That way we can identify if its the first restore request as attrs will not be set by then.

shreyanshjain7174 replied:
@soumyakoduri

thotz commented on src/rgw/rgw_op.cc:
```diff
@@ -5279,29 +5279,38 @@ void RGWRestoreObj::execute(optional_yield y)
+    auto iter = bl.cbegin();
+    decode(rt, iter);
+  }
+  if (attr_iter != attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::None) {
```
no need IMHO

shreyanshjain7174 replied:
if it's not there then for the first request it takes RestoreAlreadyInprogress status and throws 409.

thotz commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +465,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+  }
+  ldpp_dout(this, 20) << "expiration date: " << dump_time_to_str(expiration_date) << dendl;
+  //restore status
+  if (attr_iter != attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
```
This should be && than ||. Or we can check the status alone

shreyanshjain7174 replied:
I will go with checking the status alone.

soumyakoduri commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5224,13 +5224,7 @@ int RGWRados::restore_obj_from_cloud(RGWLCCloudTierCtx& tier_ctx,
 
   ceph::real_time restore_time = real_clock::now();
   {
-    char buf[32];
```
you missed to copy the changes of rgw_admin.cc here from the commit I gave - https://github.com/soumyakoduri/ceph/commit/37fae2f7058649d5a2f78ec9d5316c05a4d0e2bc

shreyanshjain7174 replied:
Done!

xxhdx1985126 committed:
crimson/osd/backfill_state: always go to Enqueuing when object is pushed
during Waiting

Fixes: https://tracker.ceph.com/issues/68061
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +465,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+    auto iter = bl.cbegin();
+    decode(expiration_date, iter);
+  }
+  ldpp_dout(this, 20) << "expiration date: " << dump_time_to_str(expiration_date) << dendl;
```
This log message should be moved to inside if loop above.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3521,31 +3542,30 @@ void RGWRestoreObj_ObjStore_S3::send_response()
+  auto attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
 
+  if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::None) {
+    if (rt == rgw::sal::RGWRestoreType::Temporary) {
```
if restore_status is None, why will RestoreType be set?

shreyanshjain7174 replied:
Not have to fetch in the send response in the restore status.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +465,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
     dump_header(s, "x-amz-replication-status", i->second);
   }
+
+  attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
```
you could probably club these with the checks at line:522.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +465,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+  ldpp_dout(this, 20) << "expiration date: " << dump_time_to_str(expiration_date) << dendl;
+  //restore status
+  if (attr_iter != attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
+      dump_header_if_nonempty(s, "x-amz-restore", "ongoing-request=\"false\", expiry-date=\""+ dump_time_to_str(expiration_date) +"\"");
```
What about RestoreAlreadyInProgress status?

xxhdx1985126 committed:
crimson/osd: purge strays when PGs go clean

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

xxhdx1985126 committed:
test/crimson/test_backfill: set BackfillState's log level to debug

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

xxhdx1985126 committed:
crimson/osd/backfill_state: do at least one time of replica scanning if
necessary in the Enqueuing state

Fixes: https://tracker.ceph.com/issues/68175
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

Commit:
mgr/dashboard: show non default realm sync status in rgw overview page

Currently, we just show the sync status of the default realm in rgw
overview page. This PR is to show the sync status of non-default realms
as well. Multisite sync status can be viewed for any of the active daemon
which runs in default/non-default realm.

Fixes: https://tracker.ceph.com/issues/68329

Signed-off-by: Aashish Sharma <aasharma@redhat.com>

kotreshhr committed:
client: Fix libcephfs aio metadata corruption.

Problem:
With cephfs nfs-ganesha, there were following
asserts hit while doing write on a file.

1. FAILED ceph_assert((bool)_front == (bool)_size)
2. FAILED ceph_assert(cap_refs[c] > 0)

Cause:
In aio path, the client_lock was not being held
in the internal callback after the io is done where
it's expected to be taken leading to corruption.

Fix:
Take client_lock in the callback

Fixes: https://tracker.ceph.com/issues/68146
Signed-off-by: Kotresh HR <khiremat@redhat.com>

kotreshhr committed:
qa: Add data read/write test for nfs-ganesha

Fixes: https://tracker.ceph.com/issues/68146
Signed-off-by: Kotresh HR <khiremat@redhat.com>

batrick committed:
qa: do not fail cephfs QA tests for slow bluestore ops

Fixes: https://tracker.ceph.com/issues/68283
Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

rkachach committed:
mgr/cephadm: adding config to check client cert for internal nginx
Fixes: https://tracker.ceph.com/issues/68310

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

kotreshhr committed:
client: Fix aio zerobyte file read

The following test fails when run with objectcacher
disabled.

TestClient.LlreadvLlwritevZeroBytes Failure - nonblocking.cc

ceph/src/osdc/Striper.cc: 186: FAILED ceph_assert(len > 0)

Traceback:
 ceph version Development (no_version) squid (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x125) [0x7fc0a340aafe]
 2: (ceph::register_assert_context(ceph::common::CephContext*)+0) [0x7fc0a340ad20]
 3: (Striper::file_to_extents(ceph::common::CephContext*, file_layout_t const*, ...)+0x184) [0x562727e13ab4]
 4: (Striper::file_to_extents(ceph::common::CephContext*, char const*, ...)+0x97) [0x562727e145d1]
 5: (Striper::file_to_extents(ceph::common::CephContext*, inodeno_t, ...)+0x75) [0x562727d29520]
 6: (Filer::read_trunc(inodeno_t, file_layout_t const*, snapid_t, ...)+0x61) [0x562727d66ea5]
 7: (Client::C_Read_Sync_NonBlocking::retry()+0x10c) [0x562727cd8a8e]
 8: (Client::_read(Fh*, long, unsigned long, ceph::buffer::v15_2_0::list*, Context*)+0x578) [0x562727d10cb6]
 9: (Client::_preadv_pwritev_locked(Fh*, iovec const*, int, long, bool, ...)+0x3a7) [0x562727d18159]
 10: (Client::ll_preadv_pwritev(Fh*, iovec const*, int, long, bool, ...)+0x179) [0x562727d18b99]
 11: (TestClient_LlreadvLlwritevZeroBytes_Test::TestBody()+0x592) [0x562727ca5352]
 12: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, ...)+0x1b) [0x562727d9dea3]
 13: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, ...)+0x80) [0x562727da2b26]
 14: (testing::Test::Run()+0xb4) [0x562727d927ae]
 15: (testing::TestInfo::Run()+0x104) [0x562727d92988]
 16: (testing::TestSuite::Run()+0xb2) [0x562727d92b34]
 17: (testing::internal::UnitTestImpl::RunAllTests()+0x36b) [0x562727d95303]
 18: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, ...)(), char const*)+0x1b) [0x562727d9e15f]
 19: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, ...)+0x80) [0x562727da3083]
 20: (testing::UnitTest::Run()+0x63) [0x562727d92813]
 21: (RUN_ALL_TESTS()+0x11) [0x562727c828d9]
 22: main()

The patch fixes the same.

Fixes: https://tracker.ceph.com/issues/68309
Signed-off-by: Kotresh HR <khiremat@redhat.com>

ktdreyer commented:
This PR is against a *fork* of Ceph, not the upstream https://github.com/ceph/ceph repository . Do you plan to submit it to https://github.com/ceph/ceph ?

kotreshhr committed:
test/client: Fix aio nonblocking test

The same bufferlist is used without cleaning
for multiple calls. The test 'LlreadvLlwritev'
used to fail because of it. Fixed the same.

Fixes: https://tracker.ceph.com/issues/68146
Signed-off-by: Kotresh HR <khiremat@redhat.com>

kotreshhr committed:
qa: Add libcephfs client test with objectcacher disabled

Fixes: https://tracker.ceph.com/issues/68146
Signed-off-by: Kotresh HR <khiremat@redhat.com>

cyx1231st committed:
crimson/os/seastore: adjust op names

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

cyx1231st committed:
crimson/os/seastore: refine logs of seastore.h/cc

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

cyx1231st committed:
crimson/os/seastore: workaround log linkage issue under clang14

The logs printing lambda-captured variables cannot be linked correctly
with SeaStore::Shard::repeat_with_onode() under clang14.

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

soumyakoduri commented:
> This PR is against a _fork_ of Ceph, not the upstream https://github.com/ceph/ceph repository . Do you plan to submit it to https://github.com/ceph/ceph ?

we are waiting for this feature PR (https://github.com/ceph/ceph/pull/59311/files) to get merged into upstream ceph post which all these PRs will be re-opened against ceph/ceph. Meanwhile we are using our private branches (which has all the commits) for internal review.

kotreshhr committed:
client: Fix caps_ref[c]<0 assert

When libcephfs aio tests (src/test/client) are run
with objectcacher disabled (ceph_test_client --client_oc=false),
the TestClient.LlreadvLlwritev fails and core dumps. The client
hits the assert 'caps_ref[c]<0'.

This patch fixes the same. There is no need to give out cap_ref
and take it again between multiple read because of short reads.
In some cases, the get_caps used to fail in C_Read_Sync_NonBlocking::finish
causing cap_ref to go negative when put_cap_ref is done at last in
C_Read_Finish::finish_io

Fixes: https://tracker.ceph.com/issues/68308
Signed-off-by: Kotresh HR <khiremat@redhat.com>

ktdreyer commented:
Thanks, that helps me understand the current state.

VallariAg committed:
qa/suites/nvmeof: use "latest" image of gateway and cli

Change nvmeof gateway and cli image from 1.2 to "latest".

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

yuvalif committed:
common: missing std include with GCC 14

In file included from src/rgw/driver/posix/bucket_cache.h:19,
                 from src/test/rgw/test_posix_bucket_cache.cc:4:
src/common/cohort_lru.h: In member function _void cohort::lru::TreeX<T, TTree, CLT, CEQ, K, LK>::lock()_:
src/common/cohort_lru.h:334:14: error: _for_each_ is not a member of _std_
  334 |         std::for_each(locks.begin(), locks.end(),
      |              ^~~~~~~~
src/common/cohort_lru.h: In member function _void cohort::lru::TreeX<T, TTree, CLT, CEQ, K, LK>::unlock()_:
/home/yuvalif/ceph5/src/common/cohort_lru.h:339:14: error: _for_each_ is not a member of _std_
  339 |         std::for_each(locks.begin(), locks.end(),
      |              ^~~~~~~~

Signed-off-by: Yuval Lifshitz <ylifshit@ibm.com>

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -1051,12 +1051,17 @@ int handle_cloudtier_obj(req_state* s, const DoutPrefixProvider *dpp, rgw::sal::
-      s->err.message = "restore is still in progress";
-    } else { // CloudRestored..return success
-      return 0;
+    } else if (restore_status == rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
```
Can you copy all the changes from https://github.com/ceph/ceph/commit/1c2b0c7ff162b830fd3092e707bb747b84c39c96 (including comments)

shreyanshjain7174 replied:
Copied.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,34 +5287,48 @@ void RGWRestoreObj::execute(optional_yield y)
-  rgw::sal::Attrs attrs = s->object->get_attrs();
-  auto attr_iter = attrs.find(RGW_ATTR_MANIFEST);
+  attrs = s->object->get_attrs();
+  auto attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
```
you do not need these checks now right..

shreyanshjain7174 replied:
Not for First request ?

soumyakoduri replied:
As per the new changes made in handle_cloudtier_obj(), 
op_ret = 0 means first request/restore request sent
op_ret =  1 -- IN_PROGRESS
op_ret = 2 -- CLOUD_RESTORED..

if there is any issue with those return status, please fix the same.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -535,6 +540,23 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+      decode(expiration_date, iter);
+    }
+    //restore status
+    if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
```
these checks should be inside if (Temporary) loop right at line:534.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -462,6 +464,9 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
       i != attrs.end()) {
     dump_header(s, "x-amz-replication-status", i->second);
   }
+
```
Please cleanup these extra lines.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3513,38 +3535,37 @@ int RGWRestoreObj_ObjStore_S3::get_params(optional_yield y)
-    dump_header(s, "x-amz-restore", rgw_bl_str(restore_status));
-  } else if (restore_status != rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
+  if (attr_iter == attrs.end() || restore_op == 0) {
+    if (rt == rgw::sal::RGWRestoreType::Temporary) {
```
Do we send expiry-date even  for Accepted return code? I thought its after the restore is successful

soumyakoduri replied:
As per https://docs.aws.amazon.com/AmazonS3/latest/userguide/restoring-objects.html , Restore: header should be included in response only if IN_PROGRESS or RESTORED.

Even sample response in https://docs.aws.amazon.com/AmazonS3/latest/API/API_RestoreObject.html for 202 Accepted does not contain Restore header

shreyanshjain7174 replied:
Well I guess we had discussed that and you suggested to include it in accepted that time.

soumyakoduri replied:
Oh is it.. sorry I dint recall.. anyway please address it now.

shreyanshjain7174 replied:
okay!

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3513,38 +3535,37 @@ int RGWRestoreObj_ObjStore_S3::get_params(optional_yield y)
-  {
-    set_req_state_err(s, op_ret);
+   
+  auto attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
```
no need of finding this attr with new changes.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3513,38 +3535,37 @@ int RGWRestoreObj_ObjStore_S3::get_params(optional_yield y)
-  }
+    dump_header(s, "x-amz-restore", "on-going-request=\"true\"");
+  } else if (restore_op == 2) {
+    if (rt == rgw::sal::RGWRestoreType::Temporary) {
```
In this loop (if restore_ret == 2), add check to find RestoreType.

shreyanshjain7174 replied:
I am already checking for temporary there right ?

Commit:
mgr/dashboard: fix gateways section error:”404 - Not Found RGW Daemon not found: None”

A case was missed here where we do have a default realm created but no default_zonegorup, in that case, the existing behavior should prevail and that's not being handled. If a default_realm is created but no default_zonegroup is there, weshould continue getting the keys from daemon_name = next(iter(daemon_keys))

Fixes: https://tracker.ceph.com/issues/68376

Signed-off-by: Aashish Sharma <aasharma@redhat.com>

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,32 +5298,38 @@ void RGWRestoreObj::execute(optional_yield y)
   int op_ret = s->object->get_obj_attrs(y, this);
   if (op_ret < 0) {
     ldpp_dout(this, 1) << "failed to fetch get_obj_attrs op ret = " << op_ret << dendl;
+    restore_op = op_ret;
```
nit: could you change the name to 'restore_ret/status' just to be more clear.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,32 +5298,38 @@ void RGWRestoreObj::execute(optional_yield y)
-      ldpp_dout(this, 20) << "not cloud tier object erroring" << dendl;
-      op_ret = -ERR_INVALID_OBJECT_STATE;
+  attrs = s->object->get_attrs();
+  auto expire_iter = attrs.find(RGW_ATTR_RESTORE_EXPIRY_DATE);
```
These are no needed now in execute(). Please move them to response and that too in the checks wherever its needed (like only in case of CloudRestored). Please cleanup all attr related changes in this routine.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,32 +5298,38 @@ void RGWRestoreObj::execute(optional_yield y)
+    op_ret = -ENOENT;
   }
+  restore_op = op_ret;
+  ldpp_dout(this, 20) << "restore_op:"<< restore_op << dendl;
```
Please reword the debug statements here and also include object names for easy debugging.

soumyakoduri commented on src/rgw/rgw_op.h:
```diff
@@ -374,7 +374,9 @@ class RGWGetObj : public RGWOp {
   ceph::real_time unmod_time;
   ceph::real_time *mod_ptr;
   ceph::real_time *unmod_ptr;
+  ceph::real_time expiration_date;
```
You can define these variables local to Response() .. no need to keep them in common struct here.

soumyakoduri commented on src/rgw/rgw_op.h:
```diff
@@ -1464,6 +1466,11 @@ class RGWPutMetadataObject : public RGWOp {
 class RGWRestoreObj : public RGWOp {
 protected:
   std::optional<uint64_t> expiry_days;
+  rgw::sal::Attrs attrs;
```
same as above.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -376,6 +376,8 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
   map<string, string>::iterator riter;
   bufferlist metadata_bl;
 
+  auto expire_iter = attrs.find(RGW_ATTR_RESTORE_EXPIRY_DATE);
```
clean up.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -450,7 +452,7 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
   dump_header_if_nonempty(s, "x-amz-version-id", version_id);
   dump_header_if_nonempty(s, "x-amz-expiration", expires);
-
+  
```
cleanup

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -527,6 +532,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+        if (attr_iter != attrs.end()) {
+          bufferlist bl = attr_iter->second;
+          auto iter = bl.cbegin();
+          decode(restore_status, iter);
```
define restore_status and expiration date and other iterators here local to this if loop but not in common structs.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -527,6 +532,24 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+          decode(expiration_date, iter);
+        }
+        //restore status
+        if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
```
why do we need to dump header if attr_iter == attrs.end()?

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3513,38 +3536,29 @@ int RGWRestoreObj_ObjStore_S3::get_params(optional_yield y)
-  } else if (restore_status != rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
+  if (restore_op == 0) {
+    s->err.http_ret = 202; // OK
+    dump_header(s, "x-amz-restore", "ongoing-request=\"false\"");
```
As mentioned in the earlier commit, this header is not needed for Accepted response.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,33 +5298,24 @@ void RGWRestoreObj::execute(optional_yield y)
+    ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
+    <<". Failing with " << op_ret << dendl;
+    if (op_ret == -ERR_INVALID_OBJECT_STATE) {
+      s->err.message = "This object was transitioned to cloud-s3 or not present";
```
s->err.message is already set in handle_cloudtier_obj() for all the error cases. IMO, we need not overwrite here again unless any error message is cryptic.

shreyanshjain7174 replied:
done! removed

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,33 +5298,24 @@ void RGWRestoreObj::execute(optional_yield y)
     }
-  } else {
-    ldpp_dout(this, 20) << " manifest not found" << dendl;
+    op_ret = -ENOENT;
```
No need to override op_ret here.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,33 +5298,24 @@ void RGWRestoreObj::execute(optional_yield y)
+  op_ret = handle_cloudtier_obj(s, this, driver, attrs, false, expiry_days, true, y);
+
+  if (op_ret < 0) {
+    ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
```
nit: can you keep entire stmt in single line.

soumyakoduri commented on src/rgw/rgw_op.cc:
```diff
@@ -5282,33 +5298,24 @@ void RGWRestoreObj::execute(optional_yield y)
-  ldpp_dout(this, 20) << "completed restore" << dendl;
+  restore_ret = op_ret;
+  ldpp_dout(this, 20) << "Restore completed of object: " << *s->object 
+  << "with op ret: " << restore_ret <<dendl;
```
nit: can you keep entire stmt in single line.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -520,13 +519,34 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
       }
     } /* checksum_mode */
     auto attr_iter = attrs.find(RGW_ATTR_RESTORE_TYPE);
+    auto expire_iter = attrs.find(RGW_ATTR_RESTORE_EXPIRY_DATE);
```
please move these 3 definitions to inside if (Temporary) loop .. i.e, line:532

soumyakoduri replied:
expire_iter, expiration_date , restore_status.

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3513,38 +3533,47 @@ int RGWRestoreObj_ObjStore_S3::get_params(optional_yield y)
+    }
+    
+    if (type_iter != attrs.end()) {
+      
```
nit:extra line

soumyakoduri reviewed (commented):
looks good. Have few minor comments though

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -527,6 +526,27 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+        //restore status
+        if (restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) {
+          dump_header_if_nonempty(s, "x-amz-restore", "ongoing-request=\"false\", expiry-date=\""+ dump_time_to_str(expiration_date) +"\"");
+        } else if (restore_status == rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
```
Haven't realized earlier.. this check when moved inside here will not work right.

if restore_status == RestoreInProgress, then restore_type will never be Temporary. And if restore_type is Temporary, for sure restore_status will always be CloudRestored. Need not check again.

So could you please move this restore_status and InProgress check outside this Temporary loop and there is no need to read/check restore_status here  again if the type is Temporary.

shreyanshjain7174 replied:
Done!

soumyakoduri reviewed (commented):
Almost done.. one final issue to address. :)

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -526,7 +525,28 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+        auto iter = bl.cbegin();
+        decode(restore_status, iter);
+      }
+      if (expire_iter != attrs.end()) {
```
nit: hmm..why is expire_iter and expiration_date moved back to outside "if (Temporary)" loop? I believe it is good practice to define and use variables only in the scope where they are necessary.

shreyanshjain7174 replied:
oops! didn't saw that, moved it with restore status accidentally. Fixing.

soumyakoduri reviewed (commented):
looks good to me.. you may open this PR against ceph:main now. Also please ensure you test all the cases (initial_request, in_progress, temporary, permanent, incorrect objects etc.. ) including for read-therough

soumyakoduri commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -526,7 +525,29 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
+        ceph::real_time expiration_date;
+        
+        if (expire_iter != attrs.end()) {
+        bufferlist bl = expire_iter->second;
```
nit: spacing inside the loop.

shreyanshjain7174 replied:
got it!

soumyakoduri reviewed (approved):
Looks great. Thanks for all the patience. Please open this PR against ceph::main now post testing all the cases.

soumyakoduri commented:
Please use this tracker - https://tracker.ceph.com/issues/68013 when you are opening PR on main branch

Commit:
mgr/dashboard: unable to edit pipe config for bucket level policy of a bucket

Fixes: https://tracker.ceph.com/issues/68387

Fixes Includes:
1) Passing additional parameter for 'user' and 'mode' as the user can be either system/dashboard or other values while creating pipe.
2) Previously while removing the src/dest bucket field, we were getting same old values on editing pipe, but now it will become '*' if empty value passed from frontend.

Signed-off-by: Naman Munet <namanmunet@li-ff83bccc-26af-11b2-a85c-a4b04bfb1003.ibm.com>

NitzanMordhai committed:
qa/suites/crimson-rados/perf: add ssh keys

cbt use ssh connection by defualt, without ssh_keys the task
won't generate public key and the cbt task will fail

Fixes: https://tracker.ceph.com/issues/68421
Signed-off-by: Nitzan Mordechai <nmordech@redhat.com>

yuvalif committed:
test/rgw/lua: use stats polling instead of sleep

this makes the test more consistent as well as faster

fixes: https://tracker.ceph.com/issues/68335

Signed-off-by: Yuval Lifshitz <ylifshit@ibm.com>

ronen-fr committed:
qa/standalone/scrub: remove TEST_recovery_scrub_2

That test does no longer match the actual requirements and
implementation of scrubbing.
It was already deactivated in
https://github.com/ceph/ceph/pull/59590. Here - it is
fully removed, mainly for the sake of backporting.

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

epuertat committed:
doc: fix email

Signed-off-by: Ernesto Puerta <37327689+epuertat@users.noreply.github.com>

VallariAg committed:
qa/workunits/nvmeof/setup_subsystem.sh: use --no-group-append

In newer version of nvmeof cli, "subsystem add" needs
this tag to ensure subsystem name is value of --subsystem.
Otherwise, in newer cli version, the gateway group is appended
at the end of the subsystem name.

This fixes the teuthology nvmeof suite (currently all jobs fails
because of this).

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

joscollin committed:
doc: update 'header get' output in cephfs-journal-tool.rst

Signed-off-by: Jos Collin <jcollin@redhat.com>

JonBailey1993 committed:
common/io_exerciser: Modify is_locked_by_me call in ceph_test_rados_io_sequence

is_locked_by_me() is a function of ceph::mutex which is only used in debug builds. By using the ceph_mutex_is_locked_by_me macro, we can neatly make sure we only run this function in debug mode, allowing compilation to no longer be affected when running in release mode.

Signed-off-by: Jon Bailey <jonathan.bailey1@ibm.com>

batrick committed:
Merge PR #60204 into main

* refs/pull/60204/head:
	doc: fix email

Reviewed-by: Patrick Donnelly <pdonnell@redhat.com>
Reviewed-by: Zac Dover <zac.dover@gmail.com>
Reviewed-by: Juan Miguel Olmo <jolmomar@redhat.com>

batrick committed:
Merge PR #60210 into main

* refs/pull/60210/head:
	doc: update 'header get' output in cephfs-journal-tool.rst

Reviewed-by: Patrick Donnelly <pdonnell@redhat.com>

cbodley committed:
Merge pull request #60211 from JonBailey1993/JonBailey1993/ceph_test_rados_io_sequence_lock_fix

common/io_exerciser: Modify is_locked_by_me call in ceph_test_rados_io_sequence

Reviewed-by: Ronen Friedman <rfriedma@redhat.com>

afreen23 committed:
Merge pull request #60050 from rhcs-dashboard/non-default-realm-sync-status

mgr/dashboard: show non default realm sync status in rgw overview page

Reviewed-by: Afreen Misbah <afreen23.git@gmail.com>

afreen23 committed:
Merge pull request #60101 from rhcs-dashboard/fix-68376-main

mgr/dashboard: fix gateways section error:”404 - Not Found RGW Daemon not found: None”

Reviewed-by: Afreen Misbah <afreen23.git@gmail.com>

yuvalif committed:
Merge pull request #60168 from yuvalif/wip-yuval-68336

test/rgw/lua: use stats polling instead of sleep

yuvalif committed:
Merge pull request #60079 from yuvalif/wip-yuval-gcc-14-test-posix

test/rgw/posix: missing std include with GCC 14

Matan-B committed:
Merge pull request #59692 from cyx1231st/wip-seastore-improve-log

crimson/os/seastore: refine seastore and transaction_manager logs, cleanups

Reviewed-by: Matan Breizman <mbreizma@redhat.com>

Matan-B committed:
Merge pull request #60161 from NitzanMordhai/wip-nitzan-crimson-perf-suites-adding-ssh-auth

qa/suites/crimson-rados/perf: add ssh keys

Reviewed-by: Matan Breizman <mbreizma@redhat.com>

vshankar committed:
Merge pull request #59987 from kotreshhr/licephfs-aio-nfsganesha

client: Fix libcephfs aio metadata corruption.

Reviewed-by: Venky Shankar <vshankar@redhat.com>

zdover23 committed:
doc/dev: remove "Stable Releases and Backports"

Remove mention of "Stable Releases and Backports" from
doc/dev/developer_guide/essentials.rst.

Co-authored-by: Laura Flores <lflores@ibm.com>
Signed-off-by: Zac Dover <zac.dover@proton.me>

batrick committed:
Merge PR #60011 into main

* refs/pull/60011/head:
	qa: do not fail cephfs QA tests for slow bluestore ops

Reviewed-by: Kotresh Hiremath Ravishankar <khiremat@redhat.com>

anthonyeleven committed:
Merge pull request #60243 from zdover23/wip-doc-2024-10-10-dev-dev-guide-essentials-backporting

doc/dev: remove "Stable Releases and Backports"

VallariAg committed:
Merge pull request #60206 from VallariAg/wip-nvmeof-fix-test

qa/suites/nvmeof: Fix broken suite by using --no-group-append tag

ljflores committed:
Merge pull request #59045 from pereman2/ignore-pg-degrade-upgrade

qa/rados/upgrade: ignore PG_DEGRADED

afreen23 committed:
Merge pull request #60147 from rhcs-dashboard/sync-policy-pipe-edit-fixes

mgr/dashboard: unable to edit pipe config for bucket level policy of bucket

Reviewed-by: Afreen Misbah <afreen23.git@gmail.com>

ceph-jenkins committed:
container/build.sh: fix arm architecture tagging

The wrong string was used for comparison, and for tagging,
so the arm64 branch and sha1 images overwrote and destroyed
the amd64 images.

Signed-off-by: Dan Mick <dmick@redhat.com>

cyx1231st committed:
Merge pull request #59720 from myoungwon/wip-fix-overlapped-write

crimson/os/seastore: fix data inconsistency during ool writes

Reviewed-by: Yingxin Cheng <yingxin.cheng@intel.com>

adk3798 committed:
Merge pull request #60028 from rkachach/fix_issue_add_internal_mtls_check

mgr/cephadm: adding config to enforce clients cert check for internal nginx (mTLS)

Reviewed-by: Adam King <adking@redhat.com>

dmick committed:
Merge pull request #60255 from dmick/wip-fix-container-arch

container/build.sh: fix arm architecture tagging

Matan-B committed:
Merge pull request #59853 from xxhdx1985126/wip-crimson-pg-purge-strays

crimson/osd: purge strays when PGs go clean

Reviewed-by: Matan Breizman <mbreizma@redhat.com>

Matan-B committed:
Merge pull request #59776 from xxhdx1985126/wip-68061

crimson/osd/backfill_state: always go to Enqueuing when object is pushed during Waiting

Reviewed-by: Samuel Just <sjust@redhat.com>

Matan-B committed:
Merge pull request #59916 from xxhdx1985126/wip-68175

crimson/osd/backfill_state: do at least one time of replica scanning if necessary in the Enqueuing state

Reviewed-by: Samuel Just <sjust@redhat.com>

ronen-fr committed:
Merge pull request #60198 from ronen-fr/wip-rf-rm-recovery-2

qa/standalone/scrub: remove TEST_recovery_scrub_2

Reviewed-by: Matan Breizman <mbreizma@redhat.com>
